aten/src/ATen/native/cpu/DistanceOpsKernel.cpp

```diff
@@ -42,95 +44,185 @@ struct PDist {
 
   template <typename F>
-  static void run_parallel(Tensor& result, const Tensor& self, const scalar_t p) {
-    const scalar_t * const self_start = self.data<scalar_t>();
-    const scalar_t * const self_end = self_start + self.numel();
-    int64_t n = self.size(0);
-    int64_t m = self.size(1);
+  static void run_parallel(
+      Tensor& result,
+      const Tensor& self,
+      const scalar_t p) {
+    const scalar_t* const self_start = self.data<scalar_t>();
+    scalar_t* const res_start = result.data<scalar_t>();
 
-    scalar_t * const res_start = result.data<scalar_t>();
-    int64_t combs = result.numel(); // n * (n - 1) / 2
+    int64_t b = self.size(0);
+    int64_t n = self.size(1);
+    int64_t d = self.size(2);
+    int64_t r = result.size(1); // n * (n - 1) / 2
+
+    double n2 = n - .5;
 
     // We conceptually iterate over tuples of (i, j, k) where i is the first
     // vector from the input, j is the second, and k is the result index. This
     // parallelizes over the range of k and infers what i and j are from the
     // value of k.
-    parallel_for(0, combs, internal::GRAIN_SIZE / (16 * m), [=](int64_t k, int64_t end) {
-      const Vec pvec(p);
-      double n2 = n - .5;
-      // The -1 accounts for floating point truncation issues
-      int64_t i = static_cast<int64_t>((n2 - std::sqrt(n2 * n2 - 2 * k - 1)));
-      int64_t j = k - n * i + i * (i + 1) / 2 + i + 1;
-
-      const scalar_t * self_i = self_start + i * m;
-      const scalar_t * self_j = self_start + j * m;
-      scalar_t * res = res_start + k;
-      const scalar_t * const res_end = res_start + end;
-
-      while (res != res_end) {
-        *res = F::finish(vec256::map2_reduce_all<scalar_t>(
-          [&pvec](Vec a, Vec b) { return F::map((a - b).abs(), pvec); },
-          F::red, self_i, self_j, m), p);
-
-        res += 1;
-        self_j += m;
-        if (self_j == self_end) {
-          self_i += m;
-          self_j = self_i + m;
-        }
-      }
-    });
+    parallel_for(
+        0,
+        b * r,
+        internal::GRAIN_SIZE / (16 * d),
+        [p, self_start, res_start, b, n, d, r, n2](int64_t start, int64_t end) {
+          const Vec pvec(p);
+
+          int64_t r_k = start % r;
+          int64_t b_l = start / r;
+          // The -1 accounts for floating point truncation issues
+          int64_t n_i =
+              static_cast<int64_t>((n2 - std::sqrt(n2 * n2 - 2 * r_k - 1)));
+          int64_t n_j = r_k - n * n_i + n_i * (n_i + 1) / 2 + n_i + 1;
+
+          const scalar_t* self_i = self_start + (b_l * n + n_i) * d;
+          const scalar_t* self_j = self_start + (b_l * n + n_j) * d;
+          const scalar_t* self_end = self_start + (b_l + 1) * n * d;
+          scalar_t* res = res_start + start;
+          const scalar_t* const res_end = res_start + end;
+
+          while (res != res_end) {
+            *res = F::finish(
+                vec256::map2_reduce_all<scalar_t>(
+                    [&pvec](Vec a, Vec b) {
```

Comment:
Instead of passing pvec, you could try passing p and creating pvec in place. Might help with inlining and in turn perf.

Reply:
I don't entirely know has this magic works. As it currently stands it's passed by reference, so only copying one word instead of four. Passing p is also one word, but then it *might* have to do the copy into four which should have the same cost. That's if it doesn't get inlined. So, definitely worth the trial.